%\VignetteIndexEntry{csaw Vignette}
%\VignetteKeyword{ChIP-Seq}
%\VignetteKeyword{differential binding}
%\VignettePackage{csaw}
\documentclass[12pt]{article}

\textwidth=6.2in
\textheight=8.5in
\oddsidemargin=0.2in
\evensidemargin=0.2in
\headheight=0in
\headsep=0in

\begin{document}

\title{Introducing the csaw package}
\author{Aaron Lun}
\date{First edition 21 June 2014\\
\vspace{6pt}
Last revised 23 June 2014}
\maketitle

The csaw package is designed for \textit{de novo} detection of differentially
bound regions from ChIP-seq data sets. It uses a sliding window approach to
count reads across the genome, and tests for significant differences within
each window using the methods in the edgeR package. Though intended for
ChIP-seq data, the methods in this package can also be applied to any type of
sequencing data where changes in genomic coverage are of interest.

The csaw package provides methods for flexible counting of single- or
paired-end reads within sorted and indexed BAM files. It implements statistical
procedures that maintain control of the false discovery rate across aggregated
windows. A range of options is also provided for normalization as well as
visualization and annotation of the results. The full user's guide is available
as part of the online documentation and can be obtained by typing:
\begin{Schunk}
\begin{Sinput}
> library(csaw)
> csawUsersGuide()
\end{Sinput}
\end{Schunk}
at the R prompt to open the user's guide in a pdf viewer.

\end{document}
