%\VignetteIndexEntry{csaw Vignette}
%\VignetteKeyword{ChIP-Seq}
%\VignetteKeyword{differential binding}
%\VignettePackage{csaw}
\documentclass[12pt]{article}

\textwidth=6.2in
\textheight=8.5in
\oddsidemargin=0.2in
\evensidemargin=0.2in
\headheight=0in
\headsep=0in

\begin{document}

\title{Introducing the \texttt{csaw} package}
\author{Aaron Lun}
\date{First edition 21 June 2014\\[0.05in]
Last revised 17 July 2015}
\maketitle

The \texttt{csaw} package is designed for the \textit{de novo} detection of differentially bound regions from ChIP-seq data. 
It uses a sliding window approach to count reads across the genome.
Each window is then tested for significant differences between libraries, using the methods in the \texttt{edgeR} package. 
\texttt{csaw} can be applied to any data set containing multiple conditions, though biological replication is strongly recommended for each condition.
While intended for ChIP-seq data, the methods in this package can also be applied to any type of sequencing data where changes in genomic coverage are of interest.

More specifically, the \texttt{csaw} package provides methods for flexible counting of single- or paired-end reads within sorted and indexed BAM files. 
It implements statistical procedures that maintain control of the false discovery rate across aggregated windows. 
Procedures for normalization of window counts between libraries are described, along with methods for independent filtering of uninteresting windows.
The full user's guide is available as part of the online documentation and can be obtained by typing:
\begin{Schunk}
\begin{Sinput}
> library(csaw)
> csawUsersGuide()
\end{Sinput}
\end{Schunk}
at the R prompt to open the user's guide in a pdf viewer.

\end{document}
